Tasks/JavaToolInstallerV0/javatoolinstaller.ts

```diff
@@ -94,22 +99,156 @@ function buildFilePath(localPathRoot: string, fileEnding: string, fileNameAndPat
+    let jdkDirectory;
+    if (fileExtension === '.dmg' && os.platform() === 'darwin') {
+        // Using set because 'includes' array method requires tsconfig option "lib": ["ES2017"]
+        const volumes: Set<string> = new Set(fs.readdirSync(VOLUMES_FOLDER));
+
+        await attach(sourceFile);
+    
+        const volumePath: string = getVolumePath(volumes);
+
+        let pkgPath: string = getPackagePath(volumePath);
+        jdkDirectory = await installPkg(pkgPath);
+
+        try {
+            await detach(volumePath);
+        }
+        catch(error) {
+            taskLib.error(error.message);
+        }
+    }
+    else if (fileExtension === '.pkg' && os.platform() === 'darwin') {
+        jdkDirectory = await installPkg(sourceFile);
+    }
+    else {
+        const javaFilesExtractor = new JavaFilesExtractor();
+        jdkDirectory = await javaFilesExtractor.unzipJavaDownload(sourceFile, fileExtension, archiveExtractLocation);
+    }
+    return jdkDirectory;
+}
+
+/**
+ * Get the path to a folder inside the VOLUMES_FOLDER.
+ * Only for macOS.
+ * @param volumes VOLUMES_FOLDER contents before attaching a disk image.
+ */
+function getVolumePath(volumes: Set<string>): string {
+    const newVolumes: string[] = fs.readdirSync(VOLUMES_FOLDER).filter(volume => !volumes.has(volume));
+
+    if (newVolumes.length !== 1) {
+        throw new Error(taskLib.loc('UnsupportedDMGStructure'));
+    }
+    return path.join(VOLUMES_FOLDER, newVolumes[0]);
+}
+
+/**
+ * Get path to a .pkg file.
+ * Only for macOS.
+ * @param volumePath Path to the folder containing a .pkg file.
+ */
+function getPackagePath(volumePath: string): string {
+    const packages: string[] = fs.readdirSync(volumePath).filter(file => file.endsWith('.pkg'));
+
+    if (packages.length === 1) {
+        return path.join(volumePath, packages[0]);
+    } else if (packages.length === 0) {
+        throw new Error(taskLib.loc('NoPKGFile'));
+    } else {
+        throw new Error(taskLib.loc('SeveralPKGFiles'));
+    }
+}
+
+async function installPkg(pkgPath: string): Promise<string> {
+    if (!fs.existsSync(pkgPath)) {
+        throw new Error('PkgPathDoesNotExist');
+    }
+
+    console.log(taskLib.loc('InstallJDK'));
+
+    // Using set because 'includes' array method requires tsconfig option "lib": ["ES2017"]
+    const JDKs: Set<string> = new Set(fs.readdirSync(JDK_FOLDER));
+
+    await runPkgInstaller(pkgPath);
+
+    const newJDKs = fs.readdirSync(JDK_FOLDER).filter(jdkName => !JDKs.has(jdkName));
+
+    if (newJDKs.length !== 1) {
+        throw new Error(taskLib.loc('NewJDKIsNotInstalled'));
+    }
+
+    let jdkDirectory: string = path.join(JDK_FOLDER, newJDKs[0], JDK_HOME_FOLDER);
+    return jdkDirectory;
+}
+
+/**
+ * Run a tool with `sudo` on Linux and macOS.
+ * Precondition: `toolName` executable is in PATH.
+ */
+function sudo(toolName: string): ToolRunner {
```

Comment:
Could you move such functions to some separate module?

Reply:
Thanks! Added utils module.